src/proofnetwork/xjdf/data/resource.cpp

```diff
@@ -0,0 +1,215 @@
+        writer.writeEndElement();
+        return;
+    }
+
+    writer.writeStartElement(QStringLiteral("Resource"));
+    if (!d->id.isEmpty())
+        writer.writeAttribute(QStringLiteral("ID"), d->id);
+    if (d->orientation != ResourceOrientation::Rotate0Orientation)
+        writer.writeAttribute(QStringLiteral("Orientation"), resourceOrientationToString(d->orientation));
+    for (const auto &part : d->parts)
+        part->toXJdf(writer);
+    if (d->amountPool)
+        d->amountPool->toXJdf(writer);
+}
+
+ResourceSP Resource::fromXJdf(QXmlStreamReader &reader, const XJdfDocumentSP &document)
+{
+    ResourceSP resource;
+    QVector<PartSP> parts;
+    AmountPoolSP amountPool;
+    QString id;
+    ResourceOrientation orientation = ResourceOrientation::Rotate0Orientation;
+
+    if (reader.isStartElement() && reader.name() == QStringLiteral("Resource")) {
+        auto attributes = reader.attributes();
+        if (attributes.hasAttribute(QStringLiteral("ID")))
+            id = attributes.value(QStringLiteral("ID")).toString();
+        if (attributes.hasAttribute(QStringLiteral("Orientation")))
+            orientation = resourceOrientationFromString(attributes.value(QStringLiteral("Orientation")).toString());
+        reader.readNextStartElement();
+
+        while (!reader.atEnd() && !reader.hasError()) {
+            if (reader.isStartElement() && reader.name() == QStringLiteral("AmountPool")) {
+                amountPool = AmountPool::fromXJdf(reader, document);
+            } else if (reader.isStartElement() && reader.name() == QStringLiteral("Part")) {
+                auto part = Part::fromXJdf(reader, document);
+                if (part)
+                    parts << part;
+            } else if (reader.isStartElement()) {
+                auto creator = resourceCreator(reader.name().toString());
+                if (creator) {
+                    resource = creator(reader, document);
+                    reader.readNext();
+                    continue;
+                }
+            } else if (reader.isEndElement() && reader.name() == QStringLiteral("Resource")) {
+                break;
+            }
+            reader.readNext();
+        }
+    }
+
+    if (resource) {
+        resource->setId(id);
+        resource->setOrientation(orientation);
+        resource->setParts(parts);
+        resource->setAmountPool(amountPool);
+    }
+
+    return resource;
+}
+
+Resource::Resource(const QString &id) : Resource(*new ResourcePrivate, id)
+{}
+
+Resource::Resource(ResourcePrivate &dd, const QString &id) : XJdfAbstractNode(dd)
+{
+    setId(id);
+}
+
+void Resource::addResourceCreator(const QString &name,
+                                  std::function<ResourceSP(QXmlStreamReader &, const XJdfDocumentSP &)> &&creator)
+{
+    if (!ResourcePrivate::creators)
+        ResourcePrivate::creators =
+            new QMap<QString, std::function<ResourceSP(QXmlStreamReader &, const XJdfDocumentSP &)>>();
```

Comment:
let's store creators the same way as we do for migrations and initializers in CoreApplication.